nx/test/nx/optional_test.exs

```diff
@@ -72,23 +106,93 @@ defmodule Nx.OptionalTest do
     end
   end
 
-  describe "optional callbacks" do
+  defmodule DefnInspect do
+    import Nx.Defn, only: [defn: 2]
+
+    defn det(t) do
+      t
+      |> Nx.LinAlg.determinant()
+      |> Nx.sum()
+      |> inspect_expr()
+    end
+  end
+
+  describe "optional callbacks (def)" do
     test "calls the default impl if specific is not present" do
-      a = Nx.tensor([[1, 0, 0], [0, 1, 0], [0, 0, 1]], backend: NonSolveTestBackend)
-      b = Nx.tensor([1, 2, 3], backend: NonSolveTestBackend)
+      a = Nx.tensor([[1, 0, 0], [0, 1, 0], [0, 0, 1]], backend: NonCustomImplTestBackend)
+      b = Nx.tensor([1, 2, 3], backend: NonCustomImplTestBackend)
 
       Nx.LinAlg.solve(a, b)
 
       assert_receive :called_default_impl
     end
 
     test "calls the custom impl if it is present" do
-      a = Nx.tensor([[1, 0, 0], [0, 1, 0], [0, 0, 1]], backend: SolveTestBackend)
-      b = Nx.tensor([1, 2, 3], backend: SolveTestBackend)
+      a = Nx.tensor([[1, 0, 0], [0, 1, 0], [0, 0, 1]], backend: CustomImplTestBackend)
+      b = Nx.tensor([1, 2, 3], backend: CustomImplTestBackend)
 
       Nx.LinAlg.solve(a, b)
 
       assert_receive :called_custom_impl
     end
   end
+
+  describe "optional callbacks (defn)" do
+    test "calls the custom impl if it exists" do
+      a = Nx.tensor([[1, 1], [2, 1]], backend: CustomImplTestBackend)
+
+      Nx.LinAlg.determinant(a)
+
+      assert_receive :called_custom_impl
+    end
+
+    test "calls the default impl if custom does not exist" do
+      a = Nx.tensor([[1, 1], [2, 1]], backend: NonCustomImplTestBackend)
+
+      Nx.LinAlg.determinant(a)
+
+      assert_receive {:called_default_impl, :reshape}
+    end
+  end
+
+  describe "inspect" do
+    for backend <- [NonCustomImplTestBackend, CustomImplTestBackend] do
```

Comment:
The test for when the backend has the custom implementation works as expected, but it seems that the `defn` is yielding a Nx.Defn.Expr tensor instead when called from the default implementation.

This is a bug in the evaluator, right?